### PR TITLE
AHelp improvements

### PR DESCRIPTION
**Pull request by Cyberboss**
I tend to crash when receiving ahelps and then I lose chat history so I came up with this.

Adds a panel for showing, managing, and archiving active ahelps.

Give me questions, because I'm too tired to explain fully this shit

### CODE REVIEW

**Comment by AnturK** [comment]
You know, i'm kinda amazed how long we didn't add any dedicated ahelp panel.

**Comment by kevinz000** [comment]
@AnturK they're a crutch for the unworthy 
if i need this stuff i just open admin logs :P

File: code/modules/admin/verbs/adminhelp.dm [hunk 1]
```diff
@@ -71,9 +71,277 @@
...
+			title = "Active Tickets"
+		if(AHELP_CLOSED)
+			l2b = closed_tickets
+			title = "Closed Tickets"
+		if(AHELP_RESOLVED)
+			l2b = resolved_tickets
+			title = "Resolved Tickets"
+	if(!l2b)
+		return
+	var/dat = "<html><head><title>[title]</title></head>"
```
Comment (RemieRichards) [inline code/modules/admin/verbs/adminhelp.dm]: ```var/list/dat = list("blah blah standard html title nonsense")```

File: code/modules/admin/verbs/adminhelp.dm [hunk 2]
```diff
@@ -71,9 +71,277 @@
...
+	for(var/I in l2b)
+		var/datum/admin_help/AH = I
+		dat += "<span class='adminnotice'><b><font color=red>Ticket #[AH.id]</font>: <A HREF='?_src_=holder;ahelp=\ref[AH];ahelp_action=ticket'>[key_name(AH.initiator)]: [AH.original_message]</A></span>"
+		
+	usr << browse(dat, "window=ahelp_list[state];size=600x480")
```
Comment (RemieRichards) [inline code/modules/admin/verbs/adminhelp.dm]: ```dat.Join()```

File: code/modules/admin/verbs/adminhelp.dm [hunk 3]
```diff
@@ -71,9 +71,277 @@
...
+
+/datum/admin_help_tickets/Destroy()
+	for(var/I in active_tickets)
```
Comment (RemieRichards) [inline code/modules/admin/verbs/adminhelp.dm]: alright, finally sick of this boilerplate, we need a ```QDEL_LIST``` and a ```QDEL_NULL_LIST```

File: code/modules/admin/verbs/adminhelp.dm (hunk 2 above)
Comment (RemieRichards) [inline code/modules/admin/verbs/adminhelp.dm]: also browse window is icky, maybe use a browser datum instead (they're the cheap decent looking black ones, but you can override the CSS if you want too)

**Comment by Poojawa** [comment]
Oh good. I tried to port old Hippie's into modernized code, but it was pretty godawful in an attempt to bring it up to standards.

https://github.com/HippieStation/HippieStation13/pull/343/files Might be helpful bits and pieces, if you wanna peak at it Cyberboss.

**Comment by lzimann** [comment]
The only reason we haven't is because I was lazy. I started doing once but stopped.

**Comment by Cyberboss** [comment]
Literally I didn't even know which server it was from, I just see shit on
Reddit and wonder why the fuck we don't have it.

**Comment by Dorsisdwarf** [comment]
I want this so bad, since I get crashed by ahelps all the time. like constantly

File: code/modules/admin/verbs/adminhelp.dm [hunk 4]
```diff
@@ -71,9 +71,277 @@
...
+
+	GLOB.ahelp_tickets.active_tickets += src
+
+/datum/admin_help/proc/FullMonty(ref_src)
+	if(!ref_src)
+		ref_src = "\ref[src]"
+	. = ADMIN_FULLMONTY_NONAME(initiator.mob)
+	if(state == AHELP_ACTIVE)
+		. += " (<A HREF='?_src_=holder;ahelp=[ref_src];ahelp_action=reject'>REJT</A>)"
+		. += " (<A HREF='?_src_=holder;ahelp=[ref_src];ahelp_action=icissue'>IC</A>) (<A HREF='?_src_=holder;ahelp=[ref_src];ahelp_action=close'>CLOSE</A>)"\
```
Comment (RemieRichards) [inline code/modules/admin/verbs/adminhelp.dm]: bad ```\``` at eol

**Comment by AnturK** [comment]
This reminds me, put the adminhelp formatting in standard css please.

**Comment by Cyberboss** [comment]
Ehhh... I have to say, the browser datums don't look great for admin matters as compared to straight html

**Comment by Cyberboss** [comment]
untitled

**Comment by Cyberboss** [comment]
The main problem will be the blue used by admins against the black bg

**Comment by RemieRichards** [comment]
to be fair, I think you've just made a gross layout.

**Comment by Cyberboss** [comment]
In my defense, I loath frontend

**Comment by lzimann** [comment]
Yea that is pretty ugly.

**Comment by Cyberboss** [comment]
How about classic admin?

untitled

**Comment by Cyberboss** [comment]
Better
untitled

**Comment by Shadowlight213** [comment]
It's not that people were too lazy to do it. I offered to create an ahelp ticket system a few times, and every time everyone was super opposed to it.

**Comment by optimumtact** [comment]
@Shadowlight213 proof?

**Comment by lzimann** [comment]
My main problem is just the name "ticket", absolutely hate it.

**Comment by Dorsisdwarf** [comment]
I always called 'em threads or investigations

**Comment by optimumtact** [comment]
cases

**Comment by Poojawa** [comment]
pretty much every form of tech support is for a 'ticket' though, so it's not exactly snowflake.

**Comment by nfreader** [comment]
This might be out-of-scope for this PR, but how about getting these saved/logged to a DB?

**Comment by Qbopper** [comment]
found a bug?

I late joined sybil and saw this (the resolved ticket is one I just handled

sybil ticket

Doesn't look like I can resolve/close a ticket when someone has disconnected? I dunno

Otherwise I really like this, having logs for ahelps with a specific player in the window is a great touch and helps me avoid scrolling up in the godawful text log

**Comment by Cyberboss** [comment]
@Qbopper Handled in 8a39483

@nfreader What data do you want saved? The text log of the ticket? And where to?

**Comment by nfreader** [comment]
>@nfreader What data do you want saved? The text log of the ticket? And where to?

Let's table this for now until after it gets merged. I'll see if I can work up a proposal on my end.

**Comment by Cyberboss** [comment]
New irc commands:

`!smsg <server> <ckey> ticket <close|resolve|icissue|reject>`

Does what you think

**Comment by Shadowlight213** [comment]
Ugh. I don't really like how you're handling the irc commands, but I'll just change it later.

**Comment by PKPenguin321** [comment]
The Resolved Tickets and Closed Tickets windows are a mess, everything clumps together

**Comment by Cyberboss** [comment]
Already fixed, I'll get someone with remote to update it when. I get in

**Comment by Poojawa** [comment]
@nfreader Yogstation's 'ticket' system works with their SQL database to track ticket histories I believe. Least that was the appearance

**Comment by nfreader** [comment]
I will scream and shout until I don't have to create a ticket name when sending the first ahelp. Create the ticket automatically and let me rename it later.

**Comment by AnturK** [comment]
Not sure how fresh the bagil version is but 2 things that stand out on first use:
Make the clickable stat objects bigger, pixel hunting for these is no fun.
Compounding with that no typeable verb to show the ticket log.

**Comment by PKPenguin321** [comment]
something that irks me:
if you resolve a report, then they reply again with some new info and you reply again, it requires a new ticket. why not just allow the new ahelps to append to the old ticket?
also, maybe allow for reopening accidentally resolved tickets.
also one time i misclicked a fresh ahelp because chat jumped and instaresolved it, and then i couldnt reply to their ahelp because "it requires a ticket" or something. that could be fixed.

**Comment by Cyberboss** [comment]
@AnturK's issue was fixed. You can now reopen an old ticket if the client has no other ticket

**Comment by AnturK** [comment]
Might want to show the date of creation as X minutes ago. Saves on having to check current server time.

**Comment by MrStonedOne** [comment]
I'd like something that has both, because I hate not being able to translate x time ago to times/dates because I don't know if its rounding up or down.

**Pull request merged**